[BE-39] feat: 거점 날씨 정보 주기적 업데이트 스케줄러 추가

xxbeann opened the pull request:
## #️⃣연관된 이슈

> ex) #103 

## 📝작업 내용

>날씨 API 연동(위도, 경도 값을 기반으로 거점별 온도 불러오기)
Kelvin 온도 Celsisu로 변환
고도별 온도 공식에 따라 고도별로 온도 변환(온도는 모두 소수점에서 올림처리했습니다.)

### 스크린샷 (선택)

## 💬리뷰 요구사항(선택)

> 비즈니스 로직 작성해봤는데, 더 나은 방법이 있다면 추천 부탁드립니다. 현재는 애플리케이션이 실행 후 최초로 1번, 이후 매 정각마다 스케줄러가 실행됩니다. 또한, `updateAllBaseWeather`메서드는 모든 거점을 순회하며 온도를 받아오고 후처리 후 다시 저장합니다. 거점이 무수하게 많아질 경우 부하가 발생할 것으로 예상되나, 현재 27개 거점에 대해서는 1ms 안에 끝나는 모습을 보였습니다. 설령 부하가 발생한다 해도 거점이 몇 만개정도는 되어야 할 것이라 사료됩니다.

추가로, `log.info`, `log.warn`을 통하여 콘솔창에서 바로 오류를 확인할 수 있게 처리 했습니다. oauth에서는 로그를 지우셨길래 따로 처리하실 생각이 있으신지 궁급합니다. 또한, 커스텀에러를 마지막 `try-catch` 구문에만 적용하려 하는데 더 필요한 곳이 있을지 여쭤보고 싶습니다!
마지막으로 예외처리가 빠진 부분이 있는지 점검 부탁드립니다!


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **신규 기능**
  * 외부 날씨 API 연동을 통한 모든 기지의 자동 날씨 갱신(서비스 시작 시 1회 실행 및 주기적 스케줄링)
  * 기지의 고도 정보를 반영한 보정 온도 적용
  * 날씨 응답을 처리하는 전용 데이터 구조 및 변환 로직 추가

* **설정**
  * 외부 API 키, 스케줄 cron, Feign 연결/응답 타임아웃 설정 항목 추가
* **기타**
  * 스케줄링 기능 활성화(@EnableScheduling)
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

xxbeann committed:
feat: API response용 dto 생성

xxbeann committed:
feat: WeatherFeignClient 작성

xxbeann committed:
feat:내부에서 사용될  WeatherDTO 작성

xxbeann committed:
feat: JSON 응답 데이터에 맞게 WeatherRes 수정

Weather 배열 안 main 필드에 날씨명 존재, Main JSON 안 temp 필드에 온도 존재.
Weather를 배열로 받아와서 그 안에 main을 바로 불러들이고, JSON main 안에 temp를 이름 변경하지 않고 그대로 불러들였습니다.

xxbeann committed:
chore: file 삭제

xxbeann committed:
refactor: 필드명 구체화

기존에는 JSON 구조에 따라서 필드명과 객체명을 매핑하였습니다.
하지만 내부에서 가독성을 높이고자 필드명을 활용될 DTO와 밀접하게 변경하였습니다.

xxbeann committed:
refactor: WeatherRes 이름 변경

xxbeann committed:
chore: scheduler 어노테이션 추가

xxbeann committed:
chore: openweather api 키 설정

xxbeann committed:
feat: scheduler 주기 추가 & weatherFeignClient Config 추가

스케줄러 주기를 관리하는 필드 추가.
weatherFeignClient의 연결 시간 및 응답시간을 설정하여 타임아웃 예외를 발생할 수 있게 했습니다.

xxbeann committed:
feat: updateWeather 메서드 추가

xxbeann committed:
feat: 매 정각마다 updateAllBaseWeather 실행하는 스케줄러 구현

xxbeann committed:
feat: 스프링 앱 실행시 바로 updateAllBaseWeather 실행시키는 메서드 추가

기존 스케줄러는 매 정각마다 실행될 뿐, 앱 실행시 최초로 실행되지는 않았습니다.
이를 해결하기위해 이벤트 리스너를 도입해, 스프링 컨텍스트와 DB가 초기화된 후 바로 실행되도록 하였습니다.

xxbeann committed:
refactor: log 추가

xxbeann committed:
refactor: altitude 필드에 nullable false 추가

xxbeann committed:
refactor: 위도 경도 Long 타입에서 BigDecimal로 변경

xxbeann committed:
feat: 날씨 API 호출하여 데이터 가져오고 후 처리하는 비즈니스 로직 작성

updateAllBaseWeather 메서드는 모든 Base들을 순회하며 위도 경도 값을 받아와 날씨API 요청을 받아옵니다.
이후, WeatherDTO에는 날씨와 켈빈온도 값이 저장되고, 이를 값이 비어있지는 않은지 판단 후,
KelvinToCelsius, applyAltitudeCorrection 함수를 실행합니다.

xxbeann committed:
Merge branch 'dev' into feat-#39

xxbeann committed:
Merge branch 'dev' into feat-#39

xxbeann committed:
refactor: 온도 올림처리

xxbeann committed:
Merge branch 'dev' into feat-#39

xxbeann committed:
Merge branch 'dev' into feat-#39

xxbeann committed:
Merge branch 'dev' into feat-#39

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/base/domain/Base.java:
```diff
@@ -38,5 +38,11 @@ public class Base {
+    @Column(nullable = false)
     private Long altitude;
+
+    public void updateWeather(String weather, Double temperature) {
+        this.weather = weather;
+        this.temperature = temperature;
+    }
```
현재 `weather`가 `String` 형식입니다. weather의 경우에는 `enum`으로 정리해서 관리할 수 있을 것 같네요!

xxbeann replied:
ENUM으로 관리할까도 생각해봤는데, 프론트랑 소통해본 결과, 백엔드에서는 String으로 넘겨주고 프론트에서 한글로 변환해서 작업하기로 했습니다! 또한, 날씨 API 구현하기(노션 페이지) 보시면 아시겠지만 날씨 코드별로 엄청 자세하게 상태가 나뉘더라구요.
그래서 나중에는 날씨 코드 이용해서 더 자세하게 날씨 알 수 있도록 변경될 것 같아요!

Woojae-Jeong replied:
그렇네요 상당히 코드가 여러개가 있네요 확인했습니다

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/base/service/weather/WeatherService.java:
```diff
@@ -0,0 +1,73 @@
+                if (weatherRes == null) {
+                    log.warn("No response, baseId={}", base.getId());
+                    continue;
+                }
```
해당 코드는 `Optional` 객체를 활용할 수 있을 것 같습니다

xxbeann replied:
한 번 리팩토링 해보겠습니다!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/base/service/weather/WeatherService.java:
```diff
@@ -0,0 +1,73 @@
+                Double kelvinObj = WeatherDTO.temperature();
+                String weather = WeatherDTO.weather();
+
+                if (kelvinObj == null) {
+                    log.warn("No temperature for baseId={}", base.getId());
+                    continue;
+                }
+
+                if (weather == null) {
+                    log.warn("No weather for baseId={}", base.getId());
+                }
```
`null값 체크`를 `WeatherRes`에서 진행하는게 아니라 `WeatherService` 에서 진행하는 이유는 둘 중 하나라도 `null값` 이면 예외를 던지는 게 아니라 해당 `Base`에 대해서는 업데이트를 하지 않으려고 WeatherService에서 null값 체크를 진행하는 거죠?

xxbeann replied:
맞습니다. 해당 base 제외하고 다음 base 업데이트 할 수 있게 넘어갈 수 있도록 구현하려 했습니다!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/base/service/weather/WeatherService.java:
```diff
@@ -0,0 +1,73 @@
+                    continue;
+                }
+
+                if (weather == null) {
+                    log.warn("No weather for baseId={}", base.getId());
+                }
```
`weather`값이 빠진경우 업데이트는 스킵하지 않고 온도만 업데이트 하는 걸까요? 만약 맞는다면 api 응답으로 둘 중 하나만 `null`값이고, 나머지 항목은 `null`이 아닌응답으로 오는 경우가 있는지 궁금합니다.

xxbeann replied:
맞습니다!
사실 api에서 null값이 오는 걸 레퍼런스로 직접 확인해본 적은 없습니다.
다만, 모종의 이유(api 호출 불가, 혹은 api 측에서 문제 발생)들로 null값이 생길 수도 있을 것 같아 저렇게 구현하기는 했습니다!

Woojae-Jeong replied:
제 생각에는 아마 api응답으로 두 항목 중 하나가 `null값`으로 올 것 같진 않습니다. 왜냐하면 저 두 정보가 날씨를 표현하는 핵심 정보이다보니 아마 정상적인 응답으로 둘 중 하나가 `null`값으로 올 것 같진 않네요. 그래서 둘 중 하나가 `null`값이 들어있으면 해당 거점의 정보는 로그 기록하고 다음 거점으로 넘어가야 하지 않나 싶네요!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/base/service/weather/WeatherService.java:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+
+    private double kelvinToCelsius(double kelvin) {
+        return kelvin - 273.15;
+    }
+
+    private double applyAltitudeCorrection(double temperature, long altitude) {
+        return Math.ceil(temperature - (altitude * 0.0065));
+    }
```
리터럴 그대로 코드에 남기는 것보다 상수화 시키는 게 좋은 코드 습관입니다!

xxbeann replied:
아하 그렇군요 리팩토링해보겠습니다!

Woojae-Jeong reviewed (commented):
리뷰 남겼습니다!

xxbeann reviewed (commented):
코멘트 달았습니다!

xxbeann committed:
refactor: 리터럴 상수화

기존에 존재하던 리터럴 코드들을 상수로 관리합니다.

xxbeann committed:
chore: yml 파일 끝 개행 추가

Woojae-Jeong reviewed (commented):
`Optional` 객체를 활용하는 게 아직 반영이 되지 않은 것 같습니다!

xxbeann committed:
Merge branch 'dev' into feat-#39

xxbeann committed:
Merge branch 'dev' into feat-#39

xxbeann committed:
Merge branch 'dev' into feat-#39

xxbeann committed:
Merge branch 'dev' into feat-#39

xxbeann committed:
Merge branch 'dev' into feat-#39

xxbeann committed:
refactor: Optional 활용하여 WeatherRes 리팩토링

xxbeann committed:
refactor: kelvinObj, weather NULL 값일 때, 업데이트 하지 않고 통과

기존 로직에서는 weather, kelvinObj 중 둘 중 하나만 업데이트 하거나 일관성이 없는 문제가 있었음.
이를 둘 중 null 값이 하나라도 있으면 모두 업데이트 하지 않는 것으로 로직을 변경함.

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/base/service/weather/WeatherService.java:
```diff
@@ -20,6 +20,8 @@ public class WeatherService {
 
     private final WeatherFeignClient weatherFeignClient;
     private final BaseRepository baseRepository;
+    static final double KELVIN_TO_CELSIUS_OFFSET = 273.15; // [°C]
+    static final double ISA_LAPSE_RATE_C_PER_M   = 0.0065; // [°C/m] (표준대기 감률: 6.5°C/km)
```
👍

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/base/service/weather/WeatherService.java:
```diff
@@ -35,12 +36,15 @@ public void updateAllBaseWeather() {
+                if (weatherOpt.isEmpty()) {
                     log.warn("No response, baseId={}", base.getId());
                     continue;
                 }
```
`Optional.ifPresentOrElse()` 에 대해서 한번 알아보시고 가독성 향상은 사바사지만, 책임 분리는 할 수 있을 것 같습니다! 가독성은 선택이지만 책임 분리는 하셨으면 좋겠네요. 현재 `updateAllBaseWeather()` 메소드에 여러 책임이 부과되어 있습니다

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/base/service/weather/WeatherService.java:
```diff
@@ -49,14 +49,16 @@ public void updateAllBaseWeather() {
 
-                if (kelvinObj == null) {
-                    log.warn("No temperature for baseId={}", base.getId());
+                if (kelvinObj == null || weather == null) {
+                    if (kelvinObj == null) {
+                        log.warn("No temperature for baseId={}", base.getId());
+                    }
+                    if (weather == null) {
+                        log.warn("No weather for baseId={}", base.getId());
+                    }
```
이 부분도 책임 분리를 할 수 있을 것 같네요!

Woojae-Jeong reviewed (commented):
리뷰 남겼습니다!

Woojae-Jeong commented on src/main/java/com/econo_4factorial/newproject/base/service/weather/WeatherService.java:
```diff
@@ -0,0 +1,81 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class WeatherService {
+
+    private final WeatherFeignClient weatherFeignClient;
+    private final BaseRepository baseRepository;
```
말씀 드리는 걸 깜빡했는데, `하나의 레파지토리에 접근할 수 있는 것은 오로지 하나의 서비스여야만 한다` 이 부분도 참고해주시면 좋을 것 같습니다.

Woojae-Jeong reviewed (commented):
코드 보다보니 하나 놓친 부분이 있어서 리뷰 하나 더 남겼습니다~
